### PR TITLE
base: Improve react component bundling

### PR DESCRIPTION
**Pull request by dperpeet**
Also fix up the playground terminal example a bit.

### CODE REVIEW

File: pkg/base1/Makefile.am [hunk 1]
```diff
@@ -75,14 +80,9 @@ pkg/base1/term.min.js: pkg/base1/term.js
...
 	$(MIN_JS_RULE)
-pkg/base1/cockpit-components.min.js: pkg/base1/cockpit-components.js
-	$(MIN_JS_RULE)
-
-REACT_COMPONENTS = \
-	pkg/base1/cockpit-components-terminal.js
 
 pkg/base1/cockpit-components.js: $(REACT_COMPONENTS)
-	$(JSMODULE) -o pkg/base1/cockpit-components.js $(REACT_COMPONENTS)
+	$(AM_V_GEN) $(srcdir)/tools/missing $(srcdir)/tools/jsbundle $@ $^
```
Comment (larsu) [inline pkg/base1/Makefile.am]: `JSMODULE` is defined in the root make file. Why did you remove it here?

File: pkg/playground/react-patterns.js [hunk 2]
```diff
@@ -22,8 +22,8 @@ require([
     "react",
     "performance/dialog-view",
     "playground/react-demo-dialog",
-    "base1/cockpit-components",
-], function(cockpit, React, dialog_pattern, demo_dialog, components) {
+    "base1/cockpit-components-terminal",
+], function(cockpit, React, dialog_pattern, demo_dialog, cockpit_terminal) {
```
Comment (larsu) [inline pkg/playground/react-patterns.js]: Hm. The idea was that we have all components in the same namespace. Is there no way to do that with jsmodule?

**Comment by larsu** [comment]
Thanks for fixing the build system.

I don't think that the second commit needs to be this complex. The timeout doesn't add any value and we should really show the error on the page instead of the terminal. Maybe render a spinner until the terminal is shown? I'm happy to work on that, as I have to fix my other branch anyway.

File: pkg/playground/react-patterns.js (hunk 2 above)
Comment (dperpeet) [inline pkg/playground/react-patterns.js]: > Hm. The idea was that we have all components in the same namespace. Is there no way to do that with jsmodule?

Why do we want them in the same namespace? I think it's good to actually include what we need.

**Comment by dperpeet** [comment]
> I don't think that the second commit needs to be this complex. The timeout doesn't add any value and we should really show the error on the page instead of the terminal. Maybe render a spinner until the terminal is shown? I'm happy to work on that, as I have to fix my other branch anyway.

It's about 20 lines of extra js code... and promises were made for conditions like this one.

Why doesn't the timeout add any value? If the data isn't available after a certain time, we get an error instead of wondering what happened.

Regarding the error on the page: This is the playground, so I didn't want to add too much stuff. Followup?

File: pkg/playground/react-patterns.js (hunk 2 above)
Comment (larsu) [inline pkg/playground/react-patterns.js]: Oh, I thought that was what we agreed upon. Maybe I misunderstood.

I like the idea of having a cockpit component package that has everything. Makes it easier for consumers of the API.

File: pkg/base1/Makefile.am (hunk 1 above)
Comment (stefwalter) [inline pkg/base1/Makefile.am]: jsbundle is used to package multiple AMD modules together.

File: pkg/playground/react-patterns.js (hunk 2 above)
Comment (stefwalter) [inline pkg/playground/react-patterns.js]: I would suggest a more measured approach, where if AMD modules are not in use they all go into the global namespace. But if AMD is in use they go into individual modules. What do you think?

Comment (dperpeet) [inline pkg/playground/react-patterns.js]: > I would suggest a more measured approach, where if AMD modules are not in use they all go into the global namespace. But if AMD is in use they go into individual modules. What do you think?

Makes sense, but I think this would be good in a follow-up.

**Comment by larsu** [comment]
Please separate the two commits. I'm fine with the bundling one and I think it should go in as is. Thanks for fixing my mess ;)

One of the playground's roles is to serve as example on how to use cockpit's react components. I don't think it's reasonable to require this intricate code at call sites. This is not about the 20 lines, but about the complexity of them (even that commit contains two bugs: it leaks a callback and doesn't work for users which don't have a shell set).

I've started working to make this situation better by creating bug #4306 and wrapping this kind of terminal in a `UserTerminal` component for the terminal module at #4309.

> Why doesn't the timeout add any value? If the data isn't available after a certain time, we get an error instead of wondering what happened.

Sorry, that was quite harsh. What I tried (and failed) to say is that outputting things on the console is not visible enough compared to putting something on the page. I thought putting "Loading" there and never showing the terminal when an error occurs would be good enough for the playground. I agree though, we should show a meaningful error when fetching the user takes too much time.

**Comment by dperpeet** [comment]
@larsu done, and posted the terminal work as a followup
